src/core/weightedGraph.js

```diff
@@ -47,3 +47,27 @@ export function merge(ws: $ReadOnlyArray<WeightedGraph>): WeightedGraph {
   const weights = Weights.merge(ws.map((w) => w.weights));
   return {graph, weights};
 }
+
+/**
+ * Create a new WeightedGraph, in which some nodes were contracted together.
+ *
+ * This is a wrapper over `Graph.contractNodes`. No effort is made to contract
+ * the weights; thus, it is entirely possible that nodes will "lose" their
+ * associated weight during a contraction.
+ *
+ * These semantics are currently acceptable becuase `contractNodes` is used by
+ * the identity plugin to collapse user identities together, and we currently
+ * always give users a weight of 0 (as they do not mint cred). If we want to
+ * start giving users specific weights, or if we begin using contractNodes for
+ * other types of nodes in the graph, we should revisit this assumption.
```

Comment:
I believe this assumption doesn't hold up.

Currently this assumption is about a convention in the weights file we use.
In fact overriding the defaults:
https://github.com/sourcecred/sourcecred/blob/325487a3ea569e1a0d176b60f39245fb2243804d/src/plugins/github/declaration.js#L58-L64
https://github.com/sourcecred/sourcecred/blob/325487a3ea569e1a0d176b60f39245fb2243804d/src/plugins/discourse/declaration.js#L28-L34

Part of the `WeightedGraph` goal is to allow plugins to set weights. So as far as the core is concerned these weights should be treated as something we don't know the use-cases for. Each plugin will be free to come up with reasons why they _would_ use weights, maybe even dynamic weights, for users or other nodes.

Definitely believe weight contraction should be supported.

Reply:
Hmm, you're right that we can technically give users a non-zero weight, although that weight never has any effect (try changing the weight for user nodes in a cred instance, recompute, and see if scores change). From an implementation standpoint, the reason is that users always have a null timestamp, and nodes with a null timestamp never mint cred. But you're right that this is an un(der)documented detail of the algorithm that we shouldn't depend on.

However, I think this behavior is actually quite principled: minting cred to users doesn't make sense. Giving all users cred by default just encourages Sibyll attacks and mass account creation; having more user accounts is not valuable for the project in itself. Giving individual users direct cred minting violates the principle of neutrality, and shifts the project towards arguing about which users should directly have cred by fiat, rather than having users earn cred based on the value of their contributions. So one approach would be to document that user types always have a weight of 0.

Also, I don't see how to do weight contraction in a principled way. Suppose that you are contracting node A (of type AType) and node B (of type BType) into node C (of type CType). Suppose that A, AType, B, BType, C, and CType all have explicit weights. How would you contract them?